tests/python_package_test/test_sklearn.py

```diff
@@ -257,6 +257,51 @@ def test_binary_classification_with_custom_objective():
     assert ret < 0.05
 
 
+@pytest.mark.parametrize('use_weight', [True, False])
+def test_binary_classification_with_auto_early_stopping(use_weight):
+
+    X, y = load_breast_cancer(return_X_y=True)
+    n_estimators = 1000
+    gbm = lgb.LGBMClassifier(
+        n_estimators=n_estimators, random_state=42, verbose=-1, early_stopping=True
+    )
+    weight = np.full_like(y, 2) if use_weight else None
+    gbm.fit(X, y, sample_weight=weight)
+    assert bool(gbm.best_iteration_)
```

Comment:
Please make these tests stricter. Instead of just "`best_iteration_` is set", please add assertions checking that:

* `best_iteration_` is set *to the expected value* (e.g. that early stopping actually happened)
* the number of trees (`gbm.booster_.num_trees()`) is set to the expected value

For example, you're not testing here that `early_stopping_rounds` is set to `10` automatically if it's not provided.

If you need help with testing that early stopping actually happened, look around the other tests in the project or the one I've added in #6095.

Reply:
(this applies to all 3 tests you've added)